### DISCUSSION TITLE
Pass subdomain_id as input file into "ElementSubdomainIDGenerator"

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#22523] [Q&A Meshing] [answered]
Hi, I would like assign different element-wise subdomain_id to different portion of mesh. So far I would like to use "ElementSubdomainIDGenerator" which allows users to re-assign the subdomain_id. But I don't know how to pass the modified subdomain_id as a .txt file to be read by MOOSE subdomain_id (If I have many elements). Do I need to write a custom object somehow to read file and transfer the data to "std::vector<SubdomainID>" type? Thanks for any suggestions.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Would there happen to be an easier definition of the subdomains than a CSV file list?
Something like x>1 -> subdomain 2, otherwise subdomain 1 for example

We support parsed expressions

https://mooseframework.inl.gov/source/meshgenerators/ElementSubdomainIDGenerator.html
ElementSubdomainId will read a really long vector, the size of the number of elements, and assign those subdomain ids to each element in the order of their element ID

Guillaume

**Reply by chunhuizhao478** [reply]
Hi Guillaume, thanks for your reply. I noticed the parsed expression, however, we would like to create multiple embedded cracks and use CZM for each crack, that means I need to identify the elements just above and below the crack surfaces and assign different subdomain_ids to them and use "BreakMeshByBlockGenerator" to create interfaces. Parsed expression may not handle this properly I think.

**Reply by GiudGiud** [reply]
ok that s fair.

What are you using to generate the mesh?

**Reply by chunhuizhao478** [reply]
Currently I try to use Gmsh and read MSH file in MOOSE, then create interface. For the multiple cracks, probably I need to write scripts to get the information I want and seems not going to be easy, then may be I could just copy/paste all the ids to MOOSE, but it's gonna be messy.

**Reply by GiudGiud** [reply]
The solution here could be to write a new object that reads the CSV and does the block assigning. We have a gap in capability on this right now. 

Adding this kind of utility in MOOSE is fairly easy due to existing classes like the DelimitedFileReader

**Reply by chunhuizhao478** [reply]
Thanks, I could check that. Currently I use MESHIO to extract cracks information and just do copy/paste the modified sudmain ID/element ID, it works for the current need.
